### PR TITLE
docs: Document protocol between UA and PS

### PR DESCRIPTION
**Pull request by jrconlin**
In order to include documentation about the returned client codes, it's
good to have documentation about the protocol. This includes a brief
summary of the autopush protocol variant of WebPush.

closes https://github.com/mozilla-services/push-service/issues/31

@kitcambridge, @bbangert r?

### CODE REVIEW

File: docs/architecture.rst [hunk 1]
```diff
@@ -161,3 +161,156 @@ the URL makes operating the server more efficient.
...
+
+.. _protocol:
+
+User Agent Protocol Overview
+============================
+
+Autopush uses an older, modified version of the Push protocol called
+SimplePush. This protocol matches WebPush very closely, but is run over
+WebSocket connections instead of HTTP/2. (See :ref:`websocket_protocol`) This was done for several reasons,
+including need to support older services that still ran the older protocol,
```
Comment (kitcambridge) [inline docs/architecture.rst]: Did you mean "older clients"?

File: docs/architecture.rst [hunk 2]
```diff
@@ -161,3 +161,156 @@ the URL makes operating the server more efficient.
...
+as well as lack of HTTP/2 libraries that were suitable for our needs.
+
+This does mean that Autopush does not yet strictly follow the WebPush
+specification. This is a brief overview of how the protocol works.
+
+0. Connection
+-------------
+User Agents establish a WebSocket connection to a well known websocket
+URL (usually wss://push.services.mozilla.org). In Firefox, this is defined
```
Comment (kitcambridge) [inline docs/architecture.rst]: Nit: push.services.mozilla.com

File: docs/architecture.rst [hunk 3]
```diff
@@ -161,3 +161,156 @@ the URL makes operating the server more efficient.
...
+in `about:config` as `dom.push.serverURL`.
+
+Once a connection is established, a User Agent must send a "Hello" message.
+
+1. Hello
+--------
+This is a greeting command allowing a UserAgent to identify itself to the Push service.
+
+`{"messageType": "hello", "uaid": _UAID_}`
```
Comment (kitcambridge) [inline docs/architecture.rst]: Also needs `"use_webpush": true`.

File: docs/architecture.rst [hunk 4]
```diff
@@ -161,3 +161,156 @@ the URL makes operating the server more efficient.
...
+*version* is the message version identifier associated with the acknowleged subscription update.
+
+*code* is the update code corresponding with the acknowlegement
+
+ACK codes
+^^^^^^^^^
+
+    :100: Message successfully delivered to Application.
+
+    :101: Message received, but failed to decyrypt.
```
Comment (kitcambridge) [inline docs/architecture.rst]: decrypt

**Comment by kitcambridge** [comment]
I know we have existing protocol docs in http://mozilla-push-service.readthedocs.io/en/latest/design/#simplepush-protocol, but I feel like these do a better job of capturing how it works overall, while the Simple Push docs delve into implementation details.

File: docs/architecture.rst (hunk 1 above)
Comment (jrconlin) [inline docs/architecture.rst]: Actually, I'm thinking older services, like Loop/Hello, Find My Device, etc. Those were a bit more than just client work.

File: docs/architecture.rst [hunk 5]
```diff
@@ -161,3 +161,159 @@ the URL makes operating the server more efficient.
...
 of the URL. This is why it's important to store the entire length of the
 endpoint URL, rather than try and optimize in some manner.
+
+
+.. _protocol:
+
+User Agent Protocol Overview
+============================
+
+.. Remember to update https://github.com/mozilla-services/push-service/docs/design.md
```
Comment (bbangert) [inline docs/architecture.rst]: Does this get printed?

**Comment by bbangert** [comment]
Looks good to me. r+

File: docs/architecture.rst (hunk 5 above)
Comment (jrconlin) [inline docs/architecture.rst]: Nope. Tested it locally and no comments. Likewise the 
     
in .md docs.

Kinda handy, no?

File: docs/architecture.rst (hunk 1 above)
Comment (kitcambridge) [inline docs/architecture.rst]: :+1: Makes sense!

**Pull request merged**